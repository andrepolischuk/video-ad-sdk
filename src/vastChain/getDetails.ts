import {get, getFirstChild, getText, getAttribute} from '../xml'
import {ParsedXML, ParsedAd, VastChain, VastChainDetails, MediaFile} from '../types'
import getLinearCreative from '../vastSelectors/helpers/getLinearCreative'
import {
  isInline,
  getClickThrough,
  getCreativeData,
  isWrapper,
  getMediaFiles,
  getInteractiveFiles
} from '../vastSelectors'
import parseTime from '../vastSelectors/helpers/parseTime'

const getAdSystem = (ad: ParsedAd): Optional<string> => {
  const adTypeElement = getFirstChild(ad)
  const element = adTypeElement && get(adTypeElement, 'AdSystem')

  return element && getText(element)
}

const getSubElementValue = (
  parentElement: ParsedXML,
  tagName: string
): Optional<string> => {
  const element = parentElement && get(parentElement, tagName)

  return element && getText(element)
}

const getPricingElement = (ad: ParsedAd): Optional<ParsedXML> => {
  const adTypeElement = getFirstChild(ad)

  return adTypeElement && get(adTypeElement, 'Pricing')
}

interface Pricing {
  pricing?: string
  pricingCurrency?: string
  pricingModel?: string
}

const getPricing = (vastChain: VastChain): Pricing => {
  const {ad} = vastChain[0]
  const pricingElement = ad && getPricingElement(ad)

  if (pricingElement) {
    return {
      pricing: getText(pricingElement),
      pricingCurrency: getAttribute(pricingElement, 'currency'),
      pricingModel: getAttribute(pricingElement, 'model')
    }
  }

  if (vastChain.length > 1) {
    return getPricing(vastChain.slice(1))
  }

  return {}
}

interface Category {
  category?: string
  categoryAuthority?: string
}

const getCategory = (ad?: ParsedAd): Category => {
  const inLineElement = ad && get(ad, 'InLine')
  const categoryElement = inLineElement && get(inLineElement, 'Category')

  if (categoryElement) {
    return {
      category: getText(categoryElement),
      categoryAuthority: getAttribute(categoryElement, 'authority')
    }
  }

  return {}
}

const getVastVersion = (parsedVast?: ParsedXML): Optional<string> => {
  const vastElement = parsedVast && get(parsedVast, 'VAST')

  if (vastElement) {
    return getAttribute(vastElement, 'version')
  }

  return 'unknown'
}

const getAdIds = (vastChain: VastChain): string[] =>
  vastChain.map(({ad}) => ad && getAttribute(ad, 'id')).filter(Boolean)

const getAdSystems = (vastChain: VastChain): string[] =>
  vastChain.map(({ad}) => ad && getAdSystem(ad)).filter(Boolean)

const getCreatives = (vastChain: VastChain): ParsedXML[] =>
  vastChain.map(({ad}) => ad && getLinearCreative(ad)).filter(Boolean)

const getCreativeIds = (creatives: ParsedXML[]): string[] =>
  creatives
    .map((creative) => creative && getAttribute(creative, 'id'))
    .filter(Boolean)

const getCreativeAdIds = (creatives: ParsedXML[]): string[] =>
  creatives
    .map((creative) => creative && getAttribute(creative, 'adId'))
    .filter(Boolean)

const getAdTypeElement = (ad?: ParsedXML): Optional<ParsedXML> =>
  ad && getFirstChild(ad)

const getCreativeElement = (ad?: ParsedXML): Optional<ParsedXML> =>
  ad && getLinearCreative(ad)

const getLinearElement = (
  creativeElement?: ParsedXML
): Optional<ParsedXML> => creativeElement && get(creativeElement, 'Linear')

const getAdServingId = (adTypeElement?: ParsedXML): Optional<string> =>
  adTypeElement && getSubElementValue(adTypeElement, 'AdServingId')

const getAdvertiser = (adTypeElement?: ParsedXML): Optional<string> =>
  adTypeElement && getSubElementValue(adTypeElement, 'Advertiser')

const getAdTitle = (adTypeElement?: ParsedXML): Optional<string> =>
  adTypeElement && getSubElementValue(adTypeElement, 'AdTitle')

const getDescription = (adTypeElement?: ParsedXML): Optional<string> =>
  adTypeElement && getSubElementValue(adTypeElement, 'Description')

const getDuration = (linearElement?: ParsedXML): Optional<string> =>
  linearElement && getSubElementValue(linearElement, 'Duration')

const getDurationInMs = (duration?: string): Optional<number> =>
  duration ? parseTime(duration) : undefined

const getUniversalIdElement = (
  creativeElement?: ParsedXML
): Optional<ParsedXML> => creativeElement && get(creativeElement, 'UniversalAdId')

/**
 * Returns a summary of the passed {@link VastChain}.
 *
 * @param vastChain the {@link VastChain} from which we want the details.
 * @returns Returns a {@link VastChainDetails} object from the passed {@link VastChain}.
 */
const getDetails = (vastChain: VastChain): VastChainDetails => {
  const adIds = getAdIds(vastChain)
  const adSystems = getAdSystems(vastChain)
  const creatives = getCreatives(vastChain)
  const creativeIds = getCreativeIds(creatives)
  const creativeAdIds = getCreativeAdIds(creatives)

  const {ad, parsedXML, XML} = vastChain[0]
  const {pricing, pricingCurrency, pricingModel} = getPricing(vastChain)
  const {category, categoryAuthority} = getCategory(ad)

  const adTypeElement = getAdTypeElement(ad)
  const creativeElement = getCreativeElement(ad)
  const linearElement = getLinearElement(creativeElement)
  const adServingId = getAdServingId(adTypeElement)
  const vastVersion = getVastVersion(parsedXML)
  const advertiser = getAdvertiser(adTypeElement)
  const adTitle = getAdTitle(adTypeElement)
  const description = getDescription(adTypeElement)
  const duration = getDuration(linearElement)
  const durationInMs = getDurationInMs(duration)

  let adId
  let adWrapperIds: string[] = []
  let adSystem
  let adWrapperSystems: string[] = []
  let creativeId
  let adWrapperCreativeIds: string[] = []
  let creativeAdId
  let adWrapperCreativeAdIds: string[] = []
  let clickThroughUrl
  let creativeData
  let universalAdId
  let universalAdIdRegistry
  let mediaFiles: MediaFile[] = []
  let vpaid
  let skippable
  let skipOffset
  let skipOffsetInMs

  if (isInline(ad)) {
    ;[adId, ...adWrapperIds] = adIds
    ;[adSystem, ...adWrapperSystems] = adSystems
    ;[creativeId, ...adWrapperCreativeIds] = creativeIds
    ;[creativeAdId, ...adWrapperCreativeAdIds] = creativeAdIds

    clickThroughUrl = ad && getClickThrough(ad)
    creativeData = XML ? getCreativeData(XML) : undefined

    const universalIdElement = getUniversalIdElement(creativeElement)

    universalAdId = universalIdElement && getText(universalIdElement)
    universalAdIdRegistry =
      universalIdElement && getAttribute(universalIdElement, 'idRegistry')
    mediaFiles = ad && getMediaFiles(ad) || []
    vpaid = Boolean(ad && getInteractiveFiles(ad))
    skipOffset = linearElement && getAttribute(linearElement, 'skipoffset')
    skipOffsetInMs = skipOffset ? parseTime(skipOffset) : undefined
    skippable = Boolean(skipOffset)
  } else if (isWrapper(vastChain[0].ad)) {
    adWrapperIds = adIds
    adWrapperSystems = adSystems
    adWrapperCreativeIds = creativeIds
    adWrapperCreativeAdIds = creativeAdIds
  }

  return {
    adId,
    adServingId,
    adSystem,
    adTitle,
    advertiser,
    adWrapperCreativeAdIds,
    adWrapperCreativeIds,
    adWrapperIds,
    adWrapperSystems,
    category,
    categoryAuthority,
    clickThroughUrl,
    creativeAdId,
    creativeData,
    creativeId,
    description,
    duration,
    durationInMs,
    mediaFiles,
    pricing,
    pricingCurrency,
    pricingModel,
    skipOffset,
    skipOffsetInMs,
    skippable,
    universalAdId,
    universalAdIdRegistry,
    vastVersion,
    vpaid
  }
}

export default getDetails
