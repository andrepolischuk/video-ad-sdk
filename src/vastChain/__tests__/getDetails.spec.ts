import {
  vastWrapperXML,
  vastInlineXML,
  wrapperParsedXML,
  inlineParsedXML,
  wrapperAd,
  inlineAd,
  vastNoAdXML,
  noAdParsedXML

} from '../../../fixtures';
import getDetails from '../getDetails';

describe('getDetails', () => {
  let vastChain;
  let inlineVastChain;
  let noInlineVastChain;
  let emptyVastChain;

  beforeEach(() => {
    vastChain = [
      {
        ad: inlineAd,
        errorCode: null,
        parsedXML: inlineParsedXML,
        requestTag: 'https://test.example.com/vastadtaguri',
        XML: vastInlineXML
      },
      {
        ad: wrapperAd,
        errorCode: null,
        parsedXML: wrapperParsedXML,
        requestTag: 'https://test.example.com/vastadtaguri',
        XML: vastWrapperXML
      },
      {
        ad: wrapperAd,
        errorCode: null,
        parsedXML: wrapperParsedXML,
        requestTag: 'http://adtag.test.example.com',
        XML: vastWrapperXML
      }
    ];
    inlineVastChain = [vastChain[0]];
    noInlineVastChain = vastChain.slice(1);
    emptyVastChain = [
      {
        ad: undefined,
        errorCode: 900,
        parsedXML: noAdParsedXML,
        requestTag: 'http://adtag.test.example.com',
        XML: vastNoAdXML
      }
    ];
  });

  it('must return the adId and adWrapperIds', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      adId: 'abc123',
      adWrapperIds: ['4367623109', '4367623109']
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      adId: 'abc123',
      adWrapperIds: []
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      adId: undefined,
      adWrapperIds: ['4367623109', '4367623109']
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      adId: undefined,
      adWrapperIds: []
    }));
  });

  it('must return the adSystem and adWrapperSystems', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      adSystem: 'MOL',
      adWrapperSystems: ['MOL', 'MOL']
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      adSystem: 'MOL',
      adWrapperSystems: []
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      adSystem: undefined,
      adWrapperSystems: ['MOL', 'MOL']
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      adSystem: undefined,
      adWrapperSystems: []
    }));
  });

  it('must return the creativeId and adWrapperCreativeIds', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      adWrapperCreativeIds: ['wrapper123', 'wrapper123'],
      creativeId: 'inline123'
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      adWrapperCreativeIds: [],
      creativeId: 'inline123'
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      adWrapperCreativeIds: ['wrapper123', 'wrapper123'],
      creativeId: undefined
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      adWrapperCreativeIds: [],
      creativeId: undefined
    }));
  });

  it('must return the creativeAdId and adWrapperCreativeAdIds', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      adWrapperCreativeAdIds: ['wrapper123', 'wrapper123'],
      creativeAdId: 'inline123'
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      adWrapperCreativeAdIds: [],
      creativeAdId: 'inline123'
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      adWrapperCreativeAdIds: ['wrapper123', 'wrapper123'],
      creativeAdId: undefined
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      adWrapperCreativeAdIds: [],
      creativeAdId: undefined
    }));
  });

  it('must return the clickThroughUrl', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      clickThroughUrl: 'https://test.example.com/clickthrough'
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      clickThroughUrl: 'https://test.example.com/clickthrough'
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      clickThroughUrl: undefined
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      clickThroughUrl: undefined
    }));
  });

  it('must return the pricing, pricingCurrency and pricingModel', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      pricing: '25.00',
      pricingCurrency: 'EUR',
      pricingModel: 'cpm'
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      pricing: '25.00',
      pricingCurrency: 'EUR',
      pricingModel: 'cpm'
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      pricing: '25.00',
      pricingCurrency: 'USD',
      pricingModel: 'cpm'
    }));
    expect(getDetails([...emptyVastChain, ...inlineVastChain])).toEqual(expect.objectContaining({
      pricing: '25.00',
      pricingCurrency: 'EUR',
      pricingModel: 'cpm'
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      pricing: undefined,
      pricingCurrency: undefined,
      pricingModel: undefined
    }));
  });

  it('must return the category and categoryAuthority', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      category: 'Inline Video',
      categoryAuthority: 'IAB'
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      category: 'Inline Video',
      categoryAuthority: 'IAB'
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      category: undefined,
      categoryAuthority: undefined
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      category: undefined,
      categoryAuthority: undefined
    }));
  });

  it('must return the adServingId', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      adServingId: 'ADID_INSIMPLETEST_ABC123'
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      adServingId: 'ADID_INSIMPLETEST_ABC123'
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      adServingId: undefined
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      adServingId: undefined
    }));
  });

  it('must return the vastVersion', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      vastVersion: '4.0'
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      vastVersion: '4.0'
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      vastVersion: '3.0'
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      vastVersion: '4.0'
    }));

    expect(getDetails([
      {
        ad: undefined,
        errorCode: 900,
        parsedXML: '',
        requestTag: 'http://adtag.test.example.com',
        XML: vastNoAdXML
      }
    ])).toEqual(expect.objectContaining({
      vastVersion: 'unknown'
    }));
  });

  it('must return the advertiser and adTitle', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      adTitle: 'VAST 4.0 Test',
      advertiser: 'MOL',
      description: 'VAST Inline'
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      adTitle: 'VAST 4.0 Test',
      advertiser: 'MOL',
      description: 'VAST Inline'
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      adTitle: undefined,
      advertiser: undefined,
      description: undefined
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      adTitle: undefined,
      advertiser: undefined,
      description: undefined
    }));
  });

  it('must return the creativeData', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      creativeData: {
        AdParameters: 'qs=test&reddit=true',
        xmlEncoded: false
      }
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      creativeData: {
        AdParameters: 'qs=test&reddit=true',
        xmlEncoded: false
      }
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      creativeData: undefined
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      creativeData: undefined
    }));
  });

  it('must return the linear duration and durationInMs', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      duration: '00:00:30',
      durationInMs: 30000
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      duration: '00:00:30',
      durationInMs: 30000
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      duration: undefined,
      durationInMs: undefined
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      duration: undefined,
      durationInMs: undefined
    }));
  });

  it('must return the universalAdId', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      universalAdId: '8465',
      universalAdIdRegistry: 'Ad-ID'
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      universalAdId: '8465',
      universalAdIdRegistry: 'Ad-ID'
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      universalAdId: undefined,
      universalAdIdRegistry: undefined
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      universalAdId: undefined,
      universalAdIdRegistry: undefined
    }));
  });

  it('must return the media files', () => {
    expect(getDetails(vastChain).mediaFiles).toMatchSnapshot();
    expect(getDetails(inlineVastChain).mediaFiles).toMatchSnapshot();
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      mediaFiles: []
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      mediaFiles: []
    }));
  });

  it('must return vpaid flag', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      vpaid: false
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      vpaid: false
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      vpaid: undefined
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      vpaid: undefined
    }));
  });

  it('must return the skipOffset, skipOffsetInMs and skippable flag', () => {
    expect(getDetails(vastChain)).toEqual(expect.objectContaining({
      skipOffset: '00:00:05',
      skipOffsetInMs: 5000,
      skippable: true
    }));
    expect(getDetails(inlineVastChain)).toEqual(expect.objectContaining({
      skipOffset: '00:00:05',
      skipOffsetInMs: 5000,
      skippable: true
    }));
    expect(getDetails(noInlineVastChain)).toEqual(expect.objectContaining({
      skipOffset: undefined,
      skipOffsetInMs: undefined,
      skippable: undefined
    }));
    expect(getDetails(emptyVastChain)).toEqual(expect.objectContaining({
      skipOffset: undefined,
      skipOffsetInMs: undefined,
      skippable: undefined
    }));
  });
});
