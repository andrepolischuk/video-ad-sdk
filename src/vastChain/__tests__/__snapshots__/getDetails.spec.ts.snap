// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDetails must return the media files 1`] = `
[
  {
    "apiFramework": undefined,
    "bitrate": "600",
    "codec": undefined,
    "delivery": "progressive",
    "height": "1080",
    "id": undefined,
    "maintainAspectRatio": "true",
    "maxBitrate": undefined,
    "minBitrate": undefined,
    "scalable": "true",
    "src": "https://test.example.com/test1920x1080.mp4",
    "type": "video/mp4",
    "universalAdId": "8465",
    "width": "1920",
  },
  {
    "apiFramework": undefined,
    "bitrate": "600",
    "codec": undefined,
    "delivery": "progressive",
    "height": "432",
    "id": undefined,
    "maintainAspectRatio": "true",
    "maxBitrate": undefined,
    "minBitrate": undefined,
    "scalable": "true",
    "src": "https://test.example.com/test768x432.mp4",
    "type": "video/mp4",
    "universalAdId": "8465",
    "width": "768",
  },
  {
    "apiFramework": undefined,
    "bitrate": "600",
    "codec": undefined,
    "delivery": "progressive",
    "height": "362",
    "id": undefined,
    "maintainAspectRatio": "true",
    "maxBitrate": undefined,
    "minBitrate": undefined,
    "scalable": "true",
    "src": "https://test.example.com/test640x362.mp4",
    "type": "video/mp4",
    "universalAdId": "8465",
    "width": "640",
  },
]
`;

exports[`getDetails must return the media files 2`] = `
[
  {
    "apiFramework": undefined,
    "bitrate": "600",
    "codec": undefined,
    "delivery": "progressive",
    "height": "1080",
    "id": undefined,
    "maintainAspectRatio": "true",
    "maxBitrate": undefined,
    "minBitrate": undefined,
    "scalable": "true",
    "src": "https://test.example.com/test1920x1080.mp4",
    "type": "video/mp4",
    "universalAdId": "8465",
    "width": "1920",
  },
  {
    "apiFramework": undefined,
    "bitrate": "600",
    "codec": undefined,
    "delivery": "progressive",
    "height": "432",
    "id": undefined,
    "maintainAspectRatio": "true",
    "maxBitrate": undefined,
    "minBitrate": undefined,
    "scalable": "true",
    "src": "https://test.example.com/test768x432.mp4",
    "type": "video/mp4",
    "universalAdId": "8465",
    "width": "768",
  },
  {
    "apiFramework": undefined,
    "bitrate": "600",
    "codec": undefined,
    "delivery": "progressive",
    "height": "362",
    "id": undefined,
    "maintainAspectRatio": "true",
    "maxBitrate": undefined,
    "minBitrate": undefined,
    "scalable": "true",
    "src": "https://test.example.com/test640x362.mp4",
    "type": "video/mp4",
    "universalAdId": "8465",
    "width": "640",
  },
]
`;
